Fix osu!taiko editor operations not always updating the menu/visual state correctly

peppy opened the pull request:
Closes https://github.com/ppy/osu/issues/12909.

peppy committed:
Run all type and sample mutations through standardising methods

peppy committed:
Ensure `EditorBeatmap.Update` is called inside `PerformOnSelection` calls

peppy commented on osu.Game.Rulesets.Taiko/Edit/TaikoSelectionHandler.cs:
```diff
@@ -69,7 +69,11 @@ public void SetRimState(bool state)
+                if (h is Hit taikoHit)
+                {
+                    taikoHit.Type = state ? HitType.Rim : HitType.Centre;
+                    EditorBeatmap.Update(h);
```
I've added these calls in line to match the rest, but we can potentially decide that it should be implicitly run on all items in the selection. This would be safer, but potentially increase the overhead when not all items in the selection are actually being modified (though i believe this is quite rare).

smoogipoo commented on osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineBlueprintContainer.cs:
```diff
@@ -276,7 +276,11 @@ private void nudgeSelection(int amount)
+                EditorBeatmap.PerformOnSelection(h =>
+                {
+                    h.StartTime += adjustment;
+                    EditorBeatmap.Update(h);
```
What are all of these `Update()` calls doing? Seems like this is handled via `EndChange()` from `PerformOnSelection()` anyway?

peppy replied:
It's only handled if a modification is detected by `EditorBeatmap`. Check the flow of `batchPendingUpdates`, I'm pretty sure you'll find that it's required. Also see https://github.com/ppy/osu/pull/12916#discussion_r637540972 which I explain this.

smoogipoo replied:
Ah, for some reason I thought I saw an add to `batchPendingUpdates` inside there and thought this was duplicating it.

peppy committed:
Merge branch 'master' into fix-taiko-editor-sample-mutations

peppy commented:
Tests are failing due to feedback loops in the bindable logic. Have a potential solution in the works.

peppy committed:
Remove all local type update logic from `TaikoBeatmapConverter`

I believe the original goal was to keep this in the converter with the
idea that samples may not always be hard coupled to the strong/rim
states. But for now I think we can assume this coupling is going to
continue into the near future, so let's keep all the logic in
`TaikoHitObject`.

peppy committed:
Avoid bindable feedback causing overwrites

peppy commented:
Apart from fixing the conversion issue, I've also removed some duplicated specifications that were in `TaikoBeatmapConverter`. Rationale is provided in https://github.com/ppy/osu/pull/12916/commits/9223d85f37d33481fd0704349adaf19880728102.

peppy committed:
Move taiko `Type` to `Hit` and localise all bind handling
